chaincode/compute_plan_worker_state.go

```diff
@@ -0,0 +1,196 @@
+	if ComputePlanKey == "" {
+		return nil
+	}
+	cp, err := db.GetComputePlan(ComputePlanKey)
+	if err != nil {
+		return err
+	}
+	if !cp.CleanModels {
+		return nil
+	}
+
+	// Check for previously added intermediary models that might now be ready for deletion
+	wStateKey := cp.getCPWorkerStateKey(worker)
+	wState, err := db.GetCPWorkerState(wStateKey)
+	if err != nil {
+		return err
+	}
+
+	modelsUsed, modelsUnused, err := getModelsInUse(db, wState.IntermediaryModelsInUse)
+
+	wState.IntermediaryModelsInUse = modelsUsed
+
+	if len(modelsUnused) != 0 {
+		db.AddComputePlanEvent(cp.Key, cp.State.Status, modelsUnused)
+	}
+
+	children, err := getTupleChildren(db, tupleKey, false)
+	if err != nil {
+		return err
+	}
+
+	if len(children) > 0 {
+		// If a tuple has no children it's considered final and should not be
+		// listed in the index. Here we're in the other case: the tuple *does* have children.
+		wState.IntermediaryModelsInUse = append(wState.IntermediaryModelsInUse, modelKey)
+	}
+
+	return db.Put(wStateKey, wState)
+}
+
+// incrementWorkerTupleCount increases the total number of tuples for
+// a given compute plan and worker
+func (cp *ComputePlan) incrementWorkerTupleCount(db *LedgerDB, worker string) error {
+
+	// Add the worker to the list of workers, if missing
+	found := false
+	for _, w := range cp.Workers {
+		if w == worker {
+			found = true
+			break
+		}
+	}
+	if !found {
+		cp.Workers = append(cp.Workers, worker)
+	}
+
+	// Create or update the done count
+	wStateKey := cp.getCPWorkerStateKey(worker)
+	wState, err := db.GetCPWorkerState(wStateKey)
+	if err != nil {
+		return db.Add(wStateKey, ComputePlanWorkerState{TupleCount: 1})
+	}
+
+	wState.TupleCount++
+	return db.Put(wStateKey, wState)
+}
+
+// incrementWorkerDoneCount increases the count of tuples in the "done" state for
+// a given compute plan and worker
+func (cp *ComputePlan) incrementWorkerDoneCount(db *LedgerDB, worker string) error {
+	wStateKey := cp.getCPWorkerStateKey(worker)
+	wState, err := db.GetCPWorkerState(wStateKey)
+	if err != nil {
+		return nil
+	}
+	wState.DoneCount++
+	return db.Put(wStateKey, wState)
+}
+
+// getTupleCounts returns the number of tuples in the "done" state and the total number of tuples
+// for a given compute plan
+func (cp *ComputePlan) getTupleCounts(db *LedgerDB) (doneCount int, tupleCount int, err error) {
+	for _, worker := range cp.Workers {
+		wStateKey := cp.getCPWorkerStateKey(worker)
+		wState, err := db.GetCPWorkerState(wStateKey)
+		if err != nil {
+			return doneCount, tupleCount, nil
+		}
+		doneCount += wState.DoneCount
+		tupleCount += wState.TupleCount
+	}
+	return doneCount, tupleCount, nil
+}
+
+// getCPWorkerStateKey returns the worker state key for a given compute plan and worker
+func (cp *ComputePlan) getCPWorkerStateKey(worker string) string {
+	return fmt.Sprintf("computePlan~%v~doneCountByWorker~%v", cp.Key, worker)
```

Comment:
the function name is bit confusing (maybe I'm wrong) but look it's based on doneCountByWork only. Does it makes sens to call it getCPWorkerDoneKey?

Reply:
fixed in https://github.com/SubstraFoundation/substra-chaincode/pull/138/commits/e833fe06b2f2002848fb1b8a66b89c59bc18ff3d